File: C++/Machine-Learning/KMeansClustering.cpp
/*

👉 K-Means Clustering is an unsupervised learning algorithm that is used to solve the clustering problems in machine learning or data science.

👉 It groups the unlabeled dataset intp different clusters.

STEPS
=====
⇒ Initialize first k clusters.
⇒ Find distance of each data point from these k clusters, using Manhattan Distance Formula
⇒ Assign each point to the respective cluster based on their shortest distance from the cluster points.

*/

#include <iostream>  // basic input output streams
#include <vector>    // dynamic array
#include <algorithm> // find() -> find occurence of an element in a data structure
                     // *min_element() -> find minimum element in a data structure
#include <map>

using namespace std;

pair<int, int> clusterPt1, clusterPt2, clusterPt3; // initial cluster points
vector<char> initCluster;                          // dynamic array of initial cluster points' names
vector<double> pointDist;                          // dynamic array of distances between points
vector<char> coordName;                            // dynamic array of coordinate names
vector<char> cluster1, cluster2, cluster3;         // dynamic array for classified cluster points
map<int, pair<int, int>> clusterPtDist;            // a map for distances b/w clusters and points
int k = 3;                                         // k - value

class Coordinate
{
public:
    char coordName;       // coordinate name
    pair<int, int> coord; // coordinates
};

int getDistance(Coordinate *coord, pair<int, int> clusterPt, int x)
{
    int dist = abs(coord[x].coord.first - clusterPt.first) + abs(coord[x].coord.second - clusterPt.second);

    clusterPtDist.insert({dist, {clusterPt.first, clusterPt.second}}); // insert distance along with cluster

    return dist;
}

void clusteringAlgo(Coordinate *coord, int n)
{
    int minDist;

    // set initial cluster points
    for (int i = 0, j = 0; i < n; i++)
    {
        if (coord[i].coordName == initCluster[j])
        {
            ++j;
            clusterPt1.first = coord[i].coord.first;
            clusterPt1.second = coord[i].coord.second;
        }
    }

    for (int i = 0; i < n - k; i++)
    {
        if (find(initCluster.begin(), initCluster.end(), coord[i].coordName) == initCluster.end())
        {

            // find distance of ith point from the k clusters
            for (int j = 0; j < k; j++)
            {
                pointDist.push_back(getDistance(coord, clusterPt1, i));
                pointDist.push_back(getDistance(coord, clusterPt2, i));
                pointDist.push_back(getDistance(coord, clusterPt3, i));
            }

            // find closest cluster from ith point
            minDist = *min_element(pointDist.begin(), pointDist.end());

            for (auto k : clusterPtDist)
            {
                // update closest cluster point
                if (minDist == k.first)
                {
                    // first cluster closest to current data point
                    if (k.second.first == clusterPt1.first and k.second.second == clusterPt2.second)
                    {
                        // update cluster map with new classified point
                        cluster1.push_back(coord[i].coordName);

                        // update closest cluster point
                        clusterPt1.first = (clusterPt1.first + coord[i].coord.first) / 2;
                        clusterPt1.second = (clusterPt1.second + coord[i].coord.second) / 2;
                    }

                    // second cluster closest to current data point
                    else if (k.second.first == clusterPt2.first and k.second.second == clusterPt2.second)
                    {
                        // update cluster map with new classified point
                        cluster2.push_back(coord[i].coordName);

                        // update closest cluster point
                        clusterPt2.first = (clusterPt2.first + coord[i].coord.first) / 2;
                        clusterPt2.second = (clusterPt2.second + coord[i].coord.second) / 2;
                    }

                    // third cluster closest to current data point
                    else if (k.second.first == clusterPt3.first and k.second.second == clusterPt3.second)
                    {
                        // update cluster map with new classified point
                        cluster3.push_back(coord[i].coordName);

                        // update closest cluster point
                        clusterPt3.first = (clusterPt3.first + coord[i].coord.first) / 2;
                        clusterPt3.second = (clusterPt3.second + coord[i].coord.second) / 2;
                    }
                }
                coordName.clear(); // clear coordinate name's vector for next iteration
            }
        }
        pointDist.clear(); // clear point distance vector for next iteration
    }
}

// function to show clusters with classified points
void showClusters()
{
    cout << "Classified Data Points :\n";
    cout << "Cluster 1 : ";
    for (char c : cluster1)
        cout << c << " ";
    cout << "\nCluster 2 : ";
    for (char c : cluster2)
        cout << c << " ";
    cout << "\nCluster 3 : ";
    for (char c : cluster3)
        cout << c << " ";
}

// main function definition
int main()
{
    int n;   // number of data points
    char ch; // choice holding variable

    cout << "\n\t=== K MEANS CLUSTER ===";

    // input data points
    cout << "\nEnter no. of data points : ";
    cin >> n;

    Coordinate cltr[n]; // make n no. of coordinate points

    cout << "Enter data points (coordName x y) : ";
    for (int i = 0; i < n; i++)
        cin >> cltr[i].coordName >> cltr[i].coord.first >> cltr[i].coord.second;

    // choose initial cluster points
    cout << "Choose 3 clusters point among the following :\n";
    for (int i = 0; i < n; i++)
        cout << cltr[i].coordName << ". (" << cltr[i].coord.first << ", " << cltr[i].coord.second << ")" << endl;

    cout << "\nMAKE YOUR CHOICES : ";
    for (int i = 0; i < k; i++)
    {
        cin >> ch;
        initCluster.push_back(ch);
    }

    clusteringAlgo(cltr, n); // function call to classify the data points in clusters

    showClusters(); // function call to display clusters with classified data points

    return 0;
}

File: Python/Machine Learning/K Means Clustering/kMeansClustering.py
# 👉 K-Means Clustering is an unsupervised learning algorithm that is used to solve the clustering problems in machine learning or data science.

# 👉 It groups the unlabeled dataset into different clusters.

# STEPS
# =====
# ⇒ Initialize first k clusters.
# ⇒ Find distance of each data point from these k clusters, using Manhattan Distance Formula
# ⇒ Assign each point to the respective cluster based on their shortest distance from the cluster points.

initCluster = []  # list of initial cluster points' names
pointDist = []  # list of distance between points
coordName = []  # list of coordinates' names
clusterPt1, clusterPt2, clusterPt3 = [], [], []  # initial cluster points
cluster1 = []  # list for names of classified cluster 1 points
cluster2 = []  # list for names of classified cluster 2 points
cluster3 = []  # list for names of classified cluster 3 points
clusterPtDist = {}  # map for distances b/w clusters & points
k = 3  # k - value


# class with attributes coordinate name and x, y coordinates for each data point
class Coordinate:
    def __init__(self, coordName, x, y):
        self.coordName = coordName  # coordinate name
        self.x = x  # x - coordinate
        self.y = y  # y - coordinate


coordinateObjList = []  # list of objects of coordinate class


# calculate Manhattan Distance between a data point and cluster point
def getDistance(coordinateObjList, clusterPt, x):
    dist = abs(coordinateObjList[x].x - clusterPt[0]) + abs(coordinateObjList[x].y - clusterPt[1])

    clusterPtDist[dist] = clusterPt  # insert distance along with cluster

    return dist


# function for implementing clustering algorithm
def clusteringAlgo(coordinateObjList, n):
    # set initial cluster points
    j = 0
    for i in range(n):
        if coordinateObjList[i].coordName == initCluster[j]:
            j += 1
            if not clusterPt1:
                clusterPt1.append(coordinateObjList[i].x)
                clusterPt1.append(coordinateObjList[i].y)
            else:
                clusterPt1[0] = coordinateObjList[i].x
                clusterPt1[1] = coordinateObjList[i].y

    # iterates through all data points except for last k points
    for i in range(n - k):
        # checks if current coordinate name not present in initial cluster names
        if coordinateObjList[i].coordName not in initCluster:
            # find distance of ith point from the k clusters
            for j in range(k):
                pointDist.append(getDistance(coordinateObjList, clusterPt1, i))
                pointDist.append(getDistance(coordinateObjList, clusterPt1, i))
                pointDist.append(getDistance(coordinateObjList, clusterPt1, i))

            # find closest cluster from ith point
            minDist = min(pointDist)

            # iterate through the elements of the clusterPtDist dictionary (map)
            for dist, point in clusterPtDist.items():
                # checks if the minimum distance (minDist) matches the distance from the map (map key)
                if minDist == dist:
                    # If there is a match, it identifies the cluster point (point) corresponding to the minimum distance (minDist)

                    # first cluster closest to current data point
                    if point[0] == clusterPt1[0] and point[1] == clusterPt1[1]:
                        # update cluster1 list with current data point's name
                        cluster1.append(coordinateObjList[i].coordName)
                        # update cluster point with new point (calculate average)
                        clusterPt1[0] = (
                            clusterPt1[0] + coordinateObjList[i].x) / 2  # x - coordinate updation
                        clusterPt1[1] = (
                            clusterPt1[1] + coordinateObjList[i].y) / 2  # y - coordinate updation

                    # second cluster closest to current data point
                    elif point[0] == clusterPt2[0] and point[1] == clusterPt2[1]:
                        # update cluster2 list with current data point's name
                        cluster1.append(coordinateObjList[i].coordName)
                        # update cluster point with new point (calculate average)
                        clusterPt2[0] = (
                            clusterPt2[0] + coordinateObjList[i].x) / 2
                        clusterPt2[1] = (
                            clusterPt2[1] + coordinateObjList[i].y) / 2

                    # third cluster closest to current data point
                    elif point[0] == clusterPt3[0] and point[1] == clusterPt3[1]:
                        # update cluster3 list with current data point's name
                        cluster1.append(coordinateObjList[i].coordName)
                        # update cluster point with new point (calculate average)
                        clusterPt3[0] = (
                            clusterPt3[0] + coordinateObjList[i].x) / 2
                        clusterPt3[1] = (
                            clusterPt3[1] + coordinateObjList[i].y) / 2

                coordName.clear()  # clear coordinate name's vector for next iteration
        pointDist.clear()  # clear point distance vector for next iteration


# function to show clusters with classified points
def showClusters():
    print("Classified Data Points :\nCluster 1 :", end=" ")
    for c in cluster1:
        print(c, end=" ")

    print("\nCluster 2 :", end=" ")
    for c in cluster2:
        print(c, end=" ")

    print("\nCluster 3 :", end=" ")
    for c in cluster3:
        print(c, end=" ")


# main function definition
def main():
    print("\n\t=== K MEANS CLUSTER ===")

    # input no. of data points
    n = int(input("Enter no. of data points : "))

    print("Enter data points (coordName x y) :", end=" ")
    for i in range(n):
        inpString = input()  # input coordinate name and coordinates
        values = inpString.split()  # split input values
        # create an object with respective values
        coordinateObjList.append(Coordinate(values[0], int(values[1]), int(values[2])))

    # choose initial cluster points
    print("Choose 3 clusters point among the following :")
    for i in range(n):
        print("{0}. ({1}, {2})".format(
            coordinateObjList[i].coordName, coordinateObjList[i].x, coordinateObjList[i].y))

    inpString = input("\nMAKE YOUR CHOICES : ")
    values = inpString.split()
    for i in range(k):
        initCluster.append(values[i])

    # function call to classify the data points in clusters
    clusteringAlgo(coordinateObjList, n)

    # function call to display clusters with classified data points
    showClusters()


main()
